store/tikv/batch_coprocessor.go

```diff
@@ -0,0 +1,415 @@
+// Copyright 2020 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package tikv
+
+import (
+	"context"
+	"io"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"github.com/pingcap/errors"
+	"github.com/pingcap/kvproto/pkg/coprocessor"
+	"github.com/pingcap/kvproto/pkg/kvrpcpb"
+	"github.com/pingcap/kvproto/pkg/metapb"
+	"github.com/pingcap/tidb/kv"
+	"github.com/pingcap/tidb/store/tikv/tikvrpc"
+	"github.com/pingcap/tidb/util/execdetails"
+	"github.com/pingcap/tidb/util/logutil"
+	"github.com/pingcap/tidb/util/memory"
+	"go.uber.org/zap"
+)
+
+// batchCopTask comprises of multiple copTask that will send to same store.
+type batchCopTask struct {
+	storeAddr string
+	cmdType   tikvrpc.CmdType
+
+	regionTaskMap map[uint64]*copTask
+	copTasks      []copTaskAndRPCContext
+}
+
+type batchCopResponse struct {
+	pbResp *coprocessor.BatchResponse
+	detail *execdetails.ExecDetails
+
+	// batch Cop Response is yet to return startKey. So batchCop cannot retry partially.
+	startKey kv.Key
+	err      error
+	respSize int64
+	respTime time.Duration
+}
+
+// GetData implements the kv.ResultSubset GetData interface.
+func (rs *batchCopResponse) GetData() []byte {
+	return rs.pbResp.Data
+}
+
+// GetStartKey implements the kv.ResultSubset GetStartKey interface.
+func (rs *batchCopResponse) GetStartKey() kv.Key {
+	return rs.startKey
+}
+
+// GetExecDetails is unavailable currently, because TiFlash has not collected exec details for batch cop.
+// TODO: Will fix in near future.
+func (rs *batchCopResponse) GetExecDetails() *execdetails.ExecDetails {
+	return &execdetails.ExecDetails{}
+}
+
+// MemSize returns how many bytes of memory this response use
+func (rs *batchCopResponse) MemSize() int64 {
+	if rs.respSize != 0 {
+		return rs.respSize
+	}
+
+	// ignore rs.err
+	rs.respSize += int64(cap(rs.startKey))
+	if rs.detail != nil {
+		rs.respSize += int64(sizeofExecDetails)
+		if rs.detail.CommitDetail != nil {
+			rs.respSize += int64(sizeofCommitDetails)
+		}
+	}
+	if rs.pbResp != nil {
+		// Using a approximate size since it's hard to get a accurate value.
+		rs.respSize += int64(rs.pbResp.Size())
+	}
+	return rs.respSize
+}
+
+func (rs *batchCopResponse) RespTime() time.Duration {
+	return rs.respTime
+}
+
+type copTaskAndRPCContext struct {
+	task *copTask
+	ctx  *RPCContext
+}
+
+func buildBatchCopTasks(bo *Backoffer, cache *RegionCache, ranges *copRanges, req *kv.Request) ([]*batchCopTask, error) {
+	if req.StoreType != kv.TiFlash {
```

Comment:
It seems this check is unnecessary. Since https://github.com/pingcap/tidb/pull/16030/files#diff-c27388ffb48c6f6eaeefe07dd3243530R60

Reply:
ok